Add "On Code Reviews" blogpost

BasThomas opened the pull request:
This is a blog post on how I review code. I'm particularly interested in knowing if it flows well, or if some more context might be needed. Which is ironic, given the blog post, I know 😅

barbaravanaki commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+This is what I believe to be the **single most important part of a review**.
+- **I read the pull request's description _and_ ticket.** Yes, also the ticket.
+Seldom does that not help me getting a better understanding of the code. And
+understanding is key. This makes everyone involved's time a lot more valuable.
```
There's nothing wrong with this phrase, it reads totally find and like a regular conversation, I would just consider phrasing it like "this makes the time of everyone involved a lot more valuable". This is super nit, I just didn't know how formal you wanted the article to be :)

fassko replied:
I agree with @barbaravanaki

Maybe start with most important that reading ticket helps with understanding the PR not
> seldom does that not help ...

BasThomas replied:
Thanks! It was a sentence I was struggling with when I wrote it, for sure. Changed!

barbaravanaki commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+code that has been written in this pull request, I often feel asking about it
+in writing is difficult. If this happens, I either take a step away from the
+computer and try again later, or ask to go over it in person (face-to-face or
+in a video call, _not_ a phone call. Being able to see each other aids
```
I really like this point: so much of communication is nonverbal, even in regards to something as technical as code. Videos just add a certain layer of understanding you can't get even with 3 phone calls.

barbaravanaki commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+in a video call, _not_ a phone call. Being able to see each other aids
+understanding.)
+- **I ask questions.** I would call myself a curious person. If I see something
+I don't know about, or want to know more about (eg. a piece of code, business
```
e.g.*

barbaravanaki commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+logic), I ask. If the review already has many comments, I might do this
+outside the review itself.
+- **I re-read your review before submitting it.** I don't always do this, but I
+know it helps. I've seen the author ask about comments I wrote, and having a
```
This sentence makes sense, but uses 2 verb tenses. I would consider rewriting it to say "ask about comments I wrote, citing a hard time understanding what I meant." There are a dozen ways to do this one- I suggest picking the one you like that matches your voice :)

BasThomas replied:
I love this change! Thanks :)

barbaravanaki commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+- **I re-read your review before submitting it.** I don't always do this, but I
+know it helps. I've seen the author ask about comments I wrote, and having a
+hard time understanding what I meant. I feel that this has been happening less
+and less over time, though! But communication is hard.
```
This one is again super nit, and depends on what kind of tone you're trying to leave on. Currently, this reads as 'hey, I'm having some progess on this, but it's a super up hill battle and I'm going to outline some of the challenges with communication'- which may be what you're looking for!

But if you were trying to end on a more hopeful or optimistic note, I would suggest switching it around to say "Though communication is hard, I feel that this has been happening less over time!". Again, nothing grammatical here, I just want to make sure we create the right tones and emotions you want.

BasThomas replied:
Perfect, much better!

barbaravanaki commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+not "you".
+- **I review my own code.** Especially in larger pull requests, I look through
+the code before assigning a reviewer, either asking open questions to the
+reviewer, making notes and/or making some changes. Seeing the "bigger" picture
```
I got a bit confused beginning at "either": are you saying you ask questions to the reviewer before assigning, or saying you do X,Y, and Z before ever asking questions to the reviewer?

BasThomas replied:
I've omitted "either", I agree it was confusing and not adding any value.

barbaravanaki commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+
+Oh, and most of this also applies reviewing other things, like blog posts and
+documentation. For those in particular, I prefer printing the documents (so it
+is easy to make both corrections and remarks), also giving me an opportunity to
```
This is fine as is, I would just respectfully suggest starting a new sentence after remarks, and say "This also gives me x" just so the sentence isn't as long. Stylistically, it also adds a bit to that voice saying "hey, this is definitely a bonus feature".

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+Seldom does that not help me getting a better understanding of the code. And
+understanding is key. This makes everyone involved's time a lot more valuable.
+If something in the description or ticket is unclear to me, I try to reach out
+the the author before continuing a review.
```
```suggestion
the author before continuing a review.
```

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+understanding is key. This makes everyone involved's time a lot more valuable.
+If something in the description or ticket is unclear to me, I try to reach out
+the the author before continuing a review.
+- **I provide context.** If I come across something that "might better be
```
```suggestion
- **I provide context.** For instance, if I come across something that might better be
```

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+If something in the description or ticket is unclear to me, I try to reach out
+the the author before continuing a review.
+- **I provide context.** If I come across something that "might better be
+represented by a function rather than a computed property", I explain why I
```
```suggestion
represented by a function rather than a computed property, I explain why I
```

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+the topic if they want, and may be helpful for those who read the comment(s) in
+the future.
+- **I take the time to try and understand the code.** When I feel like I'm
+having a hard time grasping some of the decisions that have been made, or the
```
```suggestion
having a hard time grasping some of the decisions that have been made (or the
```

BasThomas replied:
Hm, why would you change that here?

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+the future.
+- **I take the time to try and understand the code.** When I feel like I'm
+having a hard time grasping some of the decisions that have been made, or the
+code that has been written in this pull request, I often feel asking about it
```
```suggestion
code itself I often feel asking about it
```

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+code that has been written in this pull request, I often feel asking about it
+in writing is difficult. If this happens, I either take a step away from the
+computer and try again later, or ask to go over it in person (face-to-face or
+in a video call, _not_ a phone call. Being able to see each other aids
```
```suggestion
in a video call, _not_ a phone call, nor email or even Slack). Being able to see each other aids
```

BasThomas replied:
I'm going to skip on these as I'm assuming anything "non-voice" is a no-go here anyway. 😊

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+in writing is difficult. If this happens, I either take a step away from the
+computer and try again later, or ask to go over it in person (face-to-face or
+in a video call, _not_ a phone call. Being able to see each other aids
+understanding.)
```
```suggestion
understanding).
```

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+I don't know about, or want to know more about (eg. a piece of code, business
+logic), I ask. If the review already has many comments, I might do this
+outside the review itself.
+- **I re-read your review before submitting it.** I don't always do this, but I
```
Not sure what you mean in this case; before submitting where exactly?

BasThomas replied:
Ah; I've changed this to `I re-read a review before I submit it.`
With that, I mean before I submit the review (in our case, on GitHub). I'm not sure how to better describe that without mentioning a specific product? 😕

NSMyself commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+know it helps. I've seen the author ask about comments I wrote, and having a
+hard time understanding what I meant. I feel that this has been happening less
+and less over time, though! But communication is hard.
+- **I feel responsibility for the code I review,** meaning I use "we", not
```
```suggestion
- **I feel responsible for the code I review,** meaning I use "we", not
```

fassko commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+I wanted to share some insights into how I review code, learning from others,
+helping others, and getting a better overview of the code base which ultimately
+allows me to write better[^1] code. This post is inspired by a recent
+[tweet](https://twitter.com/kristoferbaxter/status/1186290473322741760), a]
```
>a]

I assume is just 

> a

BasThomas replied:
Nice catch!

fassko commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+- **I point out the good things.** This goes for anything really; not just code
+reviews. If there's something you like or learned, let the author know. I
+believe this will make a big difference on how a review is perceived, and is a
+win-win all around.<br />
```
Remove `<br/>`

BasThomas replied:
I wanted to make sure there's a line break there 😄

fassko replied:
Hmm ok I didn’t know that you can use html tags in markdown.

BasThomas replied:
Markdown will end up just being html, so you can always do that! The same goes for the `<sub></sub>` at the bottom of the post :)

fassko commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+similar [blog post](https://techbeacon.com/app-dev-testing/10-commandments-navigating-code-reviews)
+(by [Angie Jones](https://twitter.com/techgirl1908)) and a [talk](https://www.youtube.com/watch?v=Ea8EiIPZvh0)
+on the topic (by [April Wensel](https://twitter.com/aprilwensel)) from 2018's
+try! Swift NYC.
```
Add link to the conference, maybe folks would be interested to attend it. ;)

BasThomas replied:
Yes!

fassko commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,87 @@
+Bonus: I'm a big fan of a tool we built at work to get a random reviewer for
+your code. It's called `lotto`, and picks a random reviewer outside of your
+team. It's great.
+
```
Do you have something you still haven’t tried or have seen that it works for others?

Basically showing that you are still learning, want people to reach out with ideas or suggestions.

BasThomas replied:
Good point; added!

BasThomas commented:
@barbaravanaki thanks so much for the stylistic improvements. I like them a lot. 😊

BasThomas commented:
@fassko 
> My 5 cents would be slitting up in multiple blog posts and showing with real examples.

Although I like the idea, my previous attempts at multiple-post blog entries have ended up in my `git stash`... 😬

Are there any points in particular that you feel would warrant examples?

> After this is done I will share with my team. ;)

🎉

BasThomas commented:
@NSMyself 

First and foremost, thanks for the fixes and points that needed clarification.

> 1.1 - be kind and thoughtful to others when reviewing (we all have bad days)
> 1.2 - assume everyone isn't an idiot or malevolent

Awesome; added.

> 1.3 - accept that there's no silver bullet and opinions may differ

I really like this, and although I do take this into account, I'll leave it out for now. Feel free to bring it up on Twitter after I post it? :)

> 1.4 - ask yourself if what you're commenting improves the codebase in any way or if its simply a nit

And then what? Do you feel we should aim to allow for nits in a pull request, or not?

> Try to maintain the codebase coherent (particularly important in large codebases)
> 2.1 - Is the approach the author is taking the preferred way of handling this problem? what's the preferred way?

Thanks for mentioning this. I have some more thoughts on the topic — it's something we can learn more about in our code base, although with separate teams and frameworks, I don't always see this as a problem. I do try to point it out and question it, though. I'll leave it out of the post for now :)

> 2.2 - does this PR follow your coding conventions?

I find this one tricky, but would try to mention it in the "comment" accompanying a review. A linter should already help with the small things, at least.

> 3. If a PR is longer than it should don't be afraid of asking its author to break it up into smaller chunks. Big PRs discourage thorough reviews (and PRs should be "atomic" in any case, one per feature)

Added!

> 4. Personally, I would try to add just a bit more content to some points since a couple of them could be more descriptive.

Any points in particular that stand out?

Thanks again!

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+
+- **I point out the good things.** This goes for anything really; not just code
+reviews. If there's something you like or learned, let the author know. I
+believe this will make a big difference on how a review is perceived, and is a
```
I ran this post through grammarly and it seems to dislike "difference on how" and wants to replace "on" with "in". Not sure if I would agree with "in" but it might be worth rephrasing a bit.

BasThomas replied:
Having trouble rephrasing. I think "in" is an improvement, so I changed that. :)

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+believe this will make a big difference on how a review is perceived, and is a
+win-win all around.<br />
+This is what I believe to be the **single most important part of a review**.
+- **I assume best intent.** I've written questionable code before, or have
```
```suggestion
- **I assume the best intentions.** I've written questionable code before, or have
```

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+win-win all around.<br />
+This is what I believe to be the **single most important part of a review**.
+- **I assume best intent.** I've written questionable code before, or have
+overseen mistakes. That happens. Always be kind and approach a review
```
```suggestion
overlooked mistakes. That happens. Always be kind and approach a review
```

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+This makes the time of everyone involved a lot more valuable. Seldom does this
+not help me getting a better understanding of the code. And understanding is
+key. If something in the description or ticket is unclear to me, I try to reach
+out the author before continuing a review.
```
👌 context is very important when reviewing

```suggestion
out to the author before continuing a review.
```

BasThomas replied:
Nice catch!

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+positively. 
+- **I read the pull request's description _and_ ticket.** Yes, also the ticket.
+This makes the time of everyone involved a lot more valuable. Seldom does this
+not help me getting a better understanding of the code. And understanding is
```
```suggestion
not help me gain a better understanding of the code. And understanding is
```

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+out the author before continuing a review.
+- **I provide context.** For instance, if I come across something that might
+better be represented by a function rather than a computed property, I explain
+why I would suggest that in the given case.
```
I think you could make a stronger case here. Calling out things like precedent in the code base, Swift stdlib implementations etc. Maybe also give an example of code where you would make this suggestion.

BasThomas replied:
Agreed. I've changed the example and added the context I would give. :)

> For instance, if I come across a `struct` that only
serves as a namespace, I wouldn't only suggest to use a caseless `enum` instead,
but explain why I think that is the case (no pun intended): a caseless `enum`
has a non-accessible `init` by default, while a `struct` can still be
initialized without a `private init`.

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+why I would suggest that in the given case.
+- **I provide references.** Whether it's a blog post, piece of code, or
+documentation I reference, I link it. This allows the author to dig deeper into
+the topic if they want, and may be helpful for those who read the comment(s) in
```
```suggestion
the topic if they want, and might be helpful for those who read the comment(s) in
```

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+Being able to see each other aids understanding).<br />
+An exception to this would be a pull request I feel is introducing too many
+concepts at once, in which case I'll either ask if we could break up the pull
+request in more bite-sized pieces, offering to help do so in person if need be.
```
I personally find it very difficult to break up a PR that has gotten out of hand, maybe you could provide some tips here

BasThomas replied:
I'd love to go follow up on that some time, but won't do so here. It's difficult, and certainly a case of "prevention is better than cure". It depends on how an author breaks up commits... so many factors.

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+making notes and/or making some changes. Seeing the "bigger" picture often helps
+me catch issues/improvements to be made early, hopefully making the code review
+more enjoyable in the process.
+- **I don't use words like "easy", "simply", or similar.** The fact something
```
IMO similar is fine to use and very different from the others. "Would it be possible to abstract this similarly to how we did that other thing?" or "This looks like it solves a problem similar to that other extension we have in the codebase. Would it make sense to unify?"

BasThomas replied:
Ah; I was only pointing out "easy" and "simply". "similar" is meant to indicate that list isn't exhaustive. I'll reword, as it seems confusing. :)

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+
+## On Non-Code Reviews
+
+Oh, and most of this also applies reviewing other things, like blog posts and
```
```suggestion
Oh, and most of this also applies to reviewing other things, like blog posts and
```

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+
+Oh, and most of this also applies reviewing other things, like blog posts and
+documentation. For those in particular, I prefer printing the documents (so it
+is easy to make both corrections and remarks). That also giving me the
```
```suggestion
is easy to make both corrections and remarks). That also gives me an
```

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+opportunity to find another place to work, away from my computer. Natural light
+helps!
+
+If possible, I always discuss it in person afterwards.
```
```suggestion
If possible, I always discuss it in person afterward.
```

Afterwards is a non American spelling (according to Grammarly)

BasThomas replied:
It indeed (somewhat) [seems like that](https://grammarist.com/usage/afterward-afterwards/). Today I learned.

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+
+---
+
+If you're interested in code review or proofreading of a document, I'm more
```
```suggestion
If you're interested in having me do code review or proofread a document for you, I'm more
```

donnywals commented on _posts/2019-10-27-on-code-reviews.md:
```diff
@@ -0,0 +1,102 @@
+and [Mustapha](https://twitter.com/ben_muta) for proofreading!</sub>
+
+[^1]: I suppose "better" is always subjective here, but I can take the learnings from these reviews and the insights of the current code base to help make certain decisions.
+[^2]: I find it frustrating that asking for these things myself is arguably one of my weaker points. 🤷‍♂️
```
Good post! I think too many people focus on PRs and code reviews as ways to point out mistakes to people, and too little on learning from each other, respecting each other's work and effort.

One thing that I would love to read your opinion on is what to do with "nits". Sometimes there are small changes that should be made to a PR to make it perfect / good. Like small changes to phrasing in documentation comments. I usually just add those as comments and approve the PR anyway. Curious to see what you think

BasThomas replied:
Thanks!

On nits:

First of all, "to make it perfect" is impossible. 😄

I think nits are an important part of a code review, but require (thorough) documentation on their expectations. I believe they should never block a pull request from being merged, and should be distinguishable from other comments (I know I brought up the documentation, so this is somewhat ironic, but we use 💅 as an unwritten rule to distinguish nits from other comments).

I wouldn't want a reviewer not to point out these small things, as in my experience in the worst case they make something just a bit better, and in the best case make you think about it a bit more, realizing a bigger improvement can be made.

On approving: I also tend to just "comment" on a pull request if there are many open questions/comments, rather than "requesting changes". That said, it would also be something we could document better so that its meaning is better understood by others.

fassko commented:
> @fassko
> 
> > My 5 cents would be slitting up in multiple blog posts and showing with real examples.
> 
> Although I like the idea, my previous attempts at multiple-post blog entries have ended up in my `git stash`... 😬
> 
> Are there any points in particular that you feel would warrant examples?
> 
> > After this is done I will share with my team. ;)
> 
> 🎉

I think adding examples for cases like **taking time to understand the code**, for instance if author has used approach or architecture or even unknown language need to dig deeper before writing a review. Or even providing references is a good point where example from real life would stress it even more.

I just saw from myself and other colleagues that showing examples showing it in real life not abstracting works very good. :)

BasThomas commented:
> Or even providing references is a good point where example from real life would stress it even more.

Added!

BasThomas commented:
Thanks all so much for helping out! :)
